### PR TITLE
React UI: Fix issue when changing query then time, the old query executed

### LINKED ISSUE #6405: React UI: When changing query then time, the old query is executed.
How to reproduce:

Take the new ui.

In the expression, write : time()
Click execute

Change the expression:
time()-time()
*do not click execute*

Change the execution time 

**Result**: the result value is time()

**Expected**: the result is 0

Additional note: this is a regression of the new ui

### PR DESCRIPTION
**Pull request by hooten**
This Pull Request fixes a bug in the querying behavior of the new React UI. The bug is described in the linked issue.

The root cause of the bug was the state management strategy in `ExpressionInput`. Previously, the query expression was passed into `ExpressionInput` as a prop, then stored as internal component state. When the query expression was changed, the parent components (which actually do the fetching) were not aware of the state changes. So when they performed a fetch, they sometimes used a stale query expression. Specifically, the query expression was stale whenever a query was executed without using the Execute button.

The solution I propose here is to lift the state management back into the parent components—`Panel` and `PanelList`— so that they know when the query expression changes and fetch using up-to-date values. In addition, I decouple state changes from querying so that each expression input keystroke does _not_ issue a new query. State changes that should perform queries do so in the lifecycle method `componentDidUpdate`. And every time a query is executed, the url is updated. This matches the behavior of the existing classic ui.

Fixes #6405

### CODE REVIEW

**Comment by hooten** [comment]
@roidelapluie @juliusv I'd appreciate a review when you get a chance. Thanks!

**Comment by juliusv** [comment]
@hooten Awesome, thanks, didn't even know we had this problem. Will give it a good review soon. One little thing I already spotted: switching between Graph/Table tabs leads to "No data queried yet" instead of (re)executing the query for the new view.

File: web/ui/react-app/src/Panel.tsx [hunk 1]
```diff
@@ -203,14 +193,21 @@ class Panel extends Component<PanelProps & PathPrefixProps, PanelState> {
...
+    this.executeQuery();
   };
 
   handleChangeResolution = (resolution: number | null) => {
     this.setOptions({ resolution: resolution });
+    this.executeQuery();
+  };
+
+  handleChangeType = (type: PanelType) => {
+    this.setOptions({ type: type });
```
Comment (juliusv) [inline web/ui/react-app/src/Panel.tsx]: I guess this is where another `this.executeQuery();` is needed as well.

Comment (hooten) [reply web/ui/react-app/src/Panel.tsx]: Good call. I will work on this

Comment (hooten) [reply web/ui/react-app/src/Panel.tsx]: @juliusv Switching between Graph/Table views now re-executes the query. I had to change a few things around. I think it's better now anyway.

File: web/ui/react-app/src/pages/PanelList.test.tsx [hunk 2]
```diff
@@ -37,4 +39,12 @@ describe('PanelList', () => {
     expect(btn.prop('color')).toEqual('primary');
     expect(btn.children().text()).toEqual('Add Panel');
   });
+
+  it('updatesURL when a query is executed', () => {
```
Comment (juliusv) [inline web/ui/react-app/src/pages/PanelList.test.tsx]: "updatesURL" -> "updates URL"

File: web/ui/react-app/src/pages/PanelList.tsx [hunk 3]
```diff
@@ -181,7 +182,7 @@ class PanelList extends Component<RouteComponentProps & PathPrefixProps, PanelLi
         </Row>
         {panels.map(({ id, options }) => (
           <Panel
-            onExecuteQuery={this.handleQueryHistory}
+            onExecuteQuery={this.onExecuteQuery}
```
Comment (juliusv) [inline web/ui/react-app/src/pages/PanelList.tsx]: The convention in React (at least in the docs like https://reactjs.org/docs/handling-events.html) seems to be to call the attribute for handlers `onX()`, but the actual function implementation it points to `handleX()`. So I think we'll want to have `handleExecuteQuery()` here.

**Comment by hooten** [comment]
Thanks for the feedback! Let me know if I should fix anything else

**Comment by juliusv** [comment]
:+1: Looks great now, thanks :)

**Comment by hooten** [comment]
Thanks for merging! Again really happy to contribute

**Pull request merged**